### ISSUE TITLE
Proposal for domain-local storage

### ISSUE THREAD

**Issue by sadiqj** [ocaml-multicore/ocaml-multicore#339]
This is a proposal for implementing domain-local storage in multicore. We've had a couple of cases where it would have been really useful recently (https://github.com/ocaml-multicore/ocaml-multicore/issues/333 and https://github.com/ocaml-multicore/ocaml-multicore/issues/336). I'm putting it here so we can get some feedback and see if there's potentially a better approach.

The current proposal would allow you to do something along the lines:

```
let dls = Domain.Local.create (fun () ->Random.State.make_self_init ())

... sometime later in a domain ...

let local_random = Domain.Local.get dls
```
To make this work we would need to:

1. Add a hash table to domain state (which would be a root)
2. Add a global domain_local_counter
3. When a new domain local is created, the domain_local_counter is incremented. This value is used as the newly created domain local's key into the domain state's hash table.
4. Retrieving the value from a domain local would involve it looking up the entry in the domain state's hash table for it's key and if not present then using the function provided to create, populate and return the value.
5. To avoid leaking entries in the domain state's hashtable we would probably need a finalizer for domain locals that removes their entries from all domain's hashtables.

(5) isn't ideal.

**Comment by stedolan** [comment]
This sounds useful, and I think you can do it more easily than this. `(Domain.self () :> int)` is always a small integer, so an array makes a fine domain-local storage table.

Detecting domain termination for finalisation does sound awkward, though. Maybe make the user do it with explicit acquire and release functions?

**Comment by sadiqj** [comment]
The necessity for the per-domain hash table is that you may have multiple domain locals (that may be in libraries you're using as well). 

What I think we need to avoid is users of domain local state needing to coordinate because that makes it almost impossible to use inside of libraries. I worry that having explicit acquire and release functions might do similar.

**Comment by lpw25** [comment]
Maybe a stupid question, but why not one array per piece of domain local state? Why isn't `Domain.Local.t` just an `'a array`? Or maybe something like `'a ref array Atomic.t` so you can enlarge the array with compare and swap if more domains appear.

**Comment by sadiqj** [comment]
That works but brings the mirror issue of (5) from earlier - unless we have a way of being notified about Domain termination we will hold on to domain local state after a domain's exited (and potentially unwittingly hand it out to a new domain).

There's an issue for potentially creating an at_exit hook for Domain (https://github.com/ocaml-multicore/ocaml-multicore/issues/326) which would solve that problem. We were hoping there might be an approach that didn't require that though.

**Comment by bluddy** [comment]
I would be a little wary of adding domain-local storage. It makes analysis of code even harder -- we're adding invisible global (usually mutable) state that's local only to the domain. Several solutions exist to the problems above (including one I just posted at #333 -- running an initial parallel_for to set up state from different threads), and the problem of false sharing in general isn't a huge one. My recommendation would be to table this proposal until it seems absolutely necessary.

**Comment by ctk21** [comment]
As an idea, why don't we export a single opaque unique OCaml object that represents the domain from the runtime? 

This object would be marked by the domain as a root. When the domain terminates, the object will be collected; then we can use the existing finalisers and weak pointers code to operate as they are. The only bit we have to be careful of is that users have no way to access the reference and keep it alive once the domain is gone.

Implementing the runtime section of this would involve:
 - add code for the new root that is: created on domain allocation, marked as a root, deregistered on domain termination.
 - add a hook which allows access to the reference to trusted code for management of the API.

This might allow us to add an API to register domain finalisers and implement weak hashmaps over domains (or weak arrays over domains) without complex runtime additions.

**Comment by kayceesrk** [comment]
> unique OCaml object 

What would be the type of this object? I like this API: https://hackage.haskell.org/package/thread-local-storage-0.1.2/docs/Data-TLS-GHC.html#v:freeAllTLS (or something similar). We don't need the explicit `free`. 

> The only bit we have to be careful of is that users have no way to access the reference and keep it alive once the domain is gone.

Rather than this, we don't we let that object be treated as any other object such that if another domain holds a reference to that object, then it is kept alive as long as it is referenced. The rest of the proposal can remain the same.

**Comment by kayceesrk** [comment]
```ocaml
module type TLS = sig
  type 'a key
  (** TLS key *)

  val new : unit -> 'a key
  (** Returns a new key for accessing thread-local variable. *)

  val set : 'a key -> 'a -> unit
  (** [set k v] updates the calling domain's thread local state to 
      associate the key [k] with value [v]. *)

  val get : 'a key -> 'a option
  (** [get k] returns [Some v] if a value was previously associated with 
      the key [k] on the calling domain. Otherwise returns [None]. *)
end
```

We can implement this with a single root in the `Caml_state`, which is reset at domain termination.

**Comment by kayceesrk** [comment]
Fixed by #372.
